### PR TITLE
Update openapi specification to create schemas for enums to mirror th…

### PR DESCRIPTION
**Pull request by lindseyburnett**
TL;DR

* update the open-api spec yaml file to define Granularity, Usage, ServiceLevel, and ProductId as enums
* update the resource classes accordingly
* enhance the EnumParamConverterProvider class to convert query params in a casing-agnostic manner
* create an interface for the db model enums to implement to give us a unified means of converting generated <-> db model enums
* unit tests!
* format things a million times because checkstyle still hates my formatter

Room for Improvements:

* miscellaneous inline comments on this PR
* There's no db.model for ProductId, but there doesn't look like there's a need for it.  Should we add it anyway for consistency with the other api enums?  My vote is no.

### CODE REVIEW

File: api/rhsm-subscriptions-api-spec.yaml [hunk 1]
```diff
@@ -654,6 +654,33 @@ components:
           schema:
             $ref: "#/components/schemas/Errors"
   schemas:
+    GranularityGenerated:
```
Comment (lindseyburnett) [inline api/rhsm-subscriptions-api-spec.yaml:657]: Opinions on name for this?  I tacked on "Generated" at the end to circumvent the competing namespace with the db.model.Granularity class.  Partly because of checkstyle and lines being too long when you fully qualify the package, as well as just general confusion having two things with the same name.  The side effect is that it shows up like this in the api-doc page now.  Thoughts? Opinions?

image

File: api/rhsm-subscriptions-api-spec.yaml [hunk 2]
```diff
@@ -654,6 +654,33 @@ components:
...
+      type: string
+      enum: [ Daily, Weekly, Monthly, Quarterly, Yearly ]
+      x-enum-varnames: [ DAILY, WEEKLY, MONTHLY, QUARTERLY, YEARLY ]
+    ServiceLevelGenerated:
+      type: string
+      enum: [ "", Premium, Standard, Self-Support, _ANY ]
```
Comment (lindseyburnett) [inline api/rhsm-subscriptions-api-spec.yaml:663]: The empty string displays like this...not a fan, but I might being too picky?

image

File: src/main/java/org/candlepin/subscriptions/db/model/EnumLookupUtil.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        }
+        else {
+            throw new IllegalArgumentException(String
+                .format("%s is not a valid value for %s", value, Granularity.class));
+        }
+    }
+
+    String getValue();
+
+    default <T extends Enum<T>> T asOpenApiEnum() {
```
Comment (lindseyburnett) [inline src/main/java/org/candlepin/subscriptions/db/model/EnumLookupUtil.java:53]: //TODO check if i still even need this anymore

File: src/main/java/org/candlepin/subscriptions/db/model/Granularity.java [hunk 4]
```diff
@@ -20,17 +20,55 @@
...
-public enum Granularity {
-    DAILY,
-    WEEKLY,
-    MONTHLY,
-    QUARTERLY,
-    YEARLY
+public enum Granularity implements EnumLookupUtil {
+    DAILY("Daily") {
+        @Override
+        public GranularityGenerated asOpenApiEnum() {
```
Comment (lindseyburnett) [inline src/main/java/org/candlepin/subscriptions/db/model/Granularity.java:35]: This is producing this warning...
image

not entiiiiiirely sure that I'm concerned about it yet?  Open to suggestions.

Note: the asOpenApiEnum returns `<T extends Enum<T>> T`

File: src/main/java/org/candlepin/subscriptions/resource/ResourceUtils.java [hunk 5]
```diff
@@ -149,19 +153,21 @@ static Pageable getPageable(Integer offset, Integer limit, Sort sort) {
...
+        if (Objects.isNull(usage)) {
             return Usage.ANY;
         }
-        Usage sanitizedUsage = Usage.fromString(usage);
-        // If the usage parameter is not one that we support, then throw an exception.
-        // If we don't, the query would default to UNSPECIFIED, which would be confusing.
-        if (StringUtils.hasLength(usage) && sanitizedUsage == Usage.UNSPECIFIED) {
+        Usage sanitizedUsage = Usage.fromString(usage.toString());
+
+        if (Objects.equals(sanitizedUsage, Usage.UNSPECIFIED) && StringUtils.hasLength(usage.toString())) {
```
Comment (lindseyburnett) [inline src/main/java/org/candlepin/subscriptions/resource/ResourceUtils.java:162]: Couple questions here

1. Is there anything downstream that is relying on a BadRequestException being thrown to handle the flow of execution?
2. Can I delete this block (as well as the sanitizeServiceLevel equivalent further down)?

If there is logic reliant on this exception being thrown, this PR will need further work.  Since the Resource.java classes are expecting a query param to be an enum now, the requests get rejected upstream, and we wouldn't have a situation where an invalid/undefined value got this far.  Which is where my second question comes from...can I delete it then.

File: api/rhsm-subscriptions-api-spec.yaml (hunk 1 above)
Comment (awood) [reply api/rhsm-subscriptions-api-spec.yaml:657]: I like `GranularityType`, `ServiceLevelType`, etc.  `Generated`.  I think putting a comment above the schema definition about wanting to avoid naming conflicts with the DB model classes is worth having also.  Otherwise, someone might be apt to shorten it in a refactor.

File: api/rhsm-subscriptions-api-spec.yaml (hunk 2 above)
Comment (awood) [reply api/rhsm-subscriptions-api-spec.yaml:663]: I'm okay with it.  Might be worth filing an issue with the SwaggerUI team though.  Or fix it, submit a PR to them, ????, profit.

File: src/main/java/org/candlepin/subscriptions/resource/ResourceUtils.java (hunk 5 above)
Comment (awood) [reply src/main/java/org/candlepin/subscriptions/resource/ResourceUtils.java:162]: 1. I don't believe anything relies on catching that exception.  Here I think we're relying on the exception bubbling up to RESTEasy which handles it.
2. I think you can delete this, yes, since the sanitization is handled via type conversion to an enum.

File: api/rhsm-subscriptions-api-spec.yaml (hunk 1 above)
Comment (lindseyburnett) [reply api/rhsm-subscriptions-api-spec.yaml:657]: I like it!

File: src/main/java/org/candlepin/subscriptions/db/model/EnumLookupUtil.java (hunk 3 above)
Comment (lindseyburnett) [reply src/main/java/org/candlepin/subscriptions/db/model/EnumLookupUtil.java:53]: Update: need the definition, don't need to default it.

File: api/rhsm-subscriptions-api-spec.yaml [hunk 6]
```diff
@@ -654,6 +654,34 @@ components:
...
+      enum: [ Daily, Weekly, Monthly, Quarterly, Yearly ]
+      x-enum-varnames: [ DAILY, WEEKLY, MONTHLY, QUARTERLY, YEARLY ]
+    ServiceLevelType:
+      type: string
+      enum: [ "", Premium, Standard, Self-Support, _ANY ]
+      x-enum-varnames: [ UNSPECIFIED, PREMIUM, STANDARD, SELF_SUPPORT, ANY ]
+    UsageType:
+      type: string
+      enum: [ "", Production, Development/Test, Disaster Recovery, _ANY ]
+      x-enum-varnames: [ UNSPECIFIED, PRODUCTION, DEVELOPMENT_TEST, DISASTER_RECOVERY, ANY ]
```
Comment (kahowell) [inline api/rhsm-subscriptions-api-spec.yaml:669]: I'd rather not maintain the same list twice; looking at the result without `x-enum-varnames`, we get var names like specified _except_ for `EMPTY` instead of `UNSPECIFIED` and `_ANY` instead of `ANY`. In spite of the inconsistency between `EMPTY` and `UNSPECIFIED`, I prefer to omit `x-enum-varnames` (in the other cases too). Perhaps we should just rename UNSPECIFIED to EMPTY in other places? Thoughts?

Comment (lindseyburnett) [reply api/rhsm-subscriptions-api-spec.yaml:669]: I'm 100% in agreement.  I actually had done that first, added x-enum-varnames because of the unspecified/empty situation, then reverted because I was trying to minimize how invasive my changes were going to be.

I didn't really like using x-enum-varnames to begin with since it's a OpenAPI vendor extension.  Personally I like to keep everything pretty vanilla as long as possible.

File: src/main/java/org/candlepin/subscriptions/resteasy/EnumParamConverterProvider.java [hunk 7]
```diff
@@ -53,34 +75,85 @@
...
+
+            Optional<Map.Entry<String, T>> result = keyValuePairs
+                .entrySet()
+                .stream()
+                .filter(kv -> value.equals(kv.getKey()))
+                .findFirst();
+
+            if (result.isPresent()) {
+                return result.get().getValue();
             }
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/resteasy/EnumParamConverterProvider.java:112]: Would be more performant (untested suggestion):
```suggestion
            T result = keyValuePairs.get(value);
            if (result != null) {
                return result;
            }
```

File: src/main/java/org/candlepin/subscriptions/resteasy/EnumParamConverterProvider.java [hunk 8]
```diff
@@ -53,34 +75,85 @@
...
+    /**
+     * ParamConverterProvider to enable use of enums in query providers in a case-insensitive manner
+     *
+     * @param <T>
+     */
+    static class CaseInsensitiveEnumParamConverter<T> extends EnumParamConverter<T>
+        implements ParamConverter<T> {
+
+        public CaseInsensitiveEnumParamConverter(Class<T> className) {
+            super(className);
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/resteasy/EnumParamConverterProvider.java:132]: Instead, I'd suggest build a map with lowercased keys and then modify `fromString` to use `map.get(value.toLowerCase())`

**Review by kahowell — CHANGES_REQUESTED** [review]
Overall, looks good, have a few suggestions for performance-sake.

File: src/main/java/org/candlepin/subscriptions/db/model/EnumLookupUtil.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+        if (match.isPresent()) {
+            return match.get();
+        }
+        else if (Objects.nonNull(defaultEnum)) {
+            return defaultEnum;
+        }
+        else {
+            throw new IllegalArgumentException(String
+                .format("%s is not a valid value for %s", value, Granularity.class));
```
[the remark this answers is not in the document]
Comment (lindseyburnett) [reply src/main/java/org/candlepin/subscriptions/db/model/EnumLookupUtil.java:47]: woops, darn rebase

File: src/main/java/org/candlepin/subscriptions/resteasy/EnumParamConverterProvider.java [hunk 10]
```diff
@@ -38,12 +48,24 @@
 @Provider
 public class EnumParamConverterProvider implements ParamConverterProvider {
 
+    public static final String INVALID_VALUE_EXCEPTION_MSG = "%s is not a valid value for %s";
+
+    protected static final Collection<Class<? extends Enum<? extends Enum<?>>>> SPECIAL_ENUMS = Arrays.asList(
```
[the remark this answers is not in the document]
Comment (lindseyburnett) [reply src/main/java/org/candlepin/subscriptions/resteasy/EnumParamConverterProvider.java:53]: The classes in this collection are the openapi generated enums, which don't implement EnumLookupUtil.  This hardcoded list has been bugging me though...I'll think on a better alternative.

**Comment by lindseyburnett** [comment]
Ready for review round 2!

Couple items I wanted to call out:

1. Originally I thought that dropping `x-enum-vars` was only going to affect `UNSPECIFIED` -> `EMPTY`.  I made those changes then realized that "any" as a type changes to... `ANY` -> `_ANY`.  Looking for everyone's blessing that I renamed that as well.
2. RE: the feedback @kahowell  had about detecting special enums via instanceof instead of maintaining an explicit list
Those generated enums don't implement the new StringValueEnum interface, and I'm not sure that it's a good idea to go mucking with the open api generator templates and figuring out how to only apply case insensitivity to some generated enums and not all.  Would probably require forking & writing a custom generator?  Combined with how we don't want to have to maintain a case-insensitive solution long term, that didn't sound very appealing.  Anyway, totally up for more recommendation/suggestions, but wanted to just give my two cents before I jump down that rabbit hole.

**Comment by mstead** [comment]
retest this please

File: src/main/java/org/candlepin/subscriptions/resteasy/EnumParamConverterProvider.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (kahowell) [reply src/main/java/org/candlepin/subscriptions/resteasy/EnumParamConverterProvider.java]: How about:  `rawType.getPackageName().startsWith("org.candlepin.subscriptions.utilization.api.model")` ? That would get everything in our API easily.

File: src/main/java/org/candlepin/subscriptions/db/model/StringValueEnum.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Interface to reduce duplication of common methods for some enums
+ */
+public interface StringValueEnum {
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/db/model/StringValueEnum.java:32]: A bit of refactoring around the generics in this class would make it a bit more readable IMO (see following suggestions).
```suggestion
public interface StringValueEnum<T> {
```

File: src/main/java/org/candlepin/subscriptions/db/model/StringValueEnum.java [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+    static <T extends Enum<T> & StringValueEnum> T getValueOf(Class<T> className, Map<String, T> valueEnumMap,
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/db/model/StringValueEnum.java:34]:
```suggestion
    static <T> T getValueOf(Class<T> className, Map<String, T> valueEnumMap,
```

File: src/main/java/org/candlepin/subscriptions/db/model/StringValueEnum.java [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+            return defaultEnum;
+        }
+        else {
+            throw new IllegalArgumentException(String
+                .format("%s is not a valid value for %s", value, className));
+        }
+
+    }
+
+    static <T extends Enum<T> & StringValueEnum> Map<String, T> initializeImmutableMap(Class<T> enumClass) {
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/db/model/StringValueEnum.java:53]:
```suggestion
    static <T extends StringValueEnum<T>> Map<String, T> initializeImmutableMap(Class<T> enumClass) {
```

File: src/main/java/org/candlepin/subscriptions/db/model/StringValueEnum.java [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+        return Arrays.stream(enumClass.getEnumConstants())
+            .collect(Collectors.toMap(t -> t.getValue().toLowerCase(), value -> value));
+    }
+
+    String getValue();
+
+    <T extends Enum<T>> T asOpenApiEnum();
```
Comment (kahowell) [inline src/main/java/org/candlepin/subscriptions/db/model/StringValueEnum.java:60]:
```suggestion
    T asOpenApiEnum();
```

**Review by kahowell — CHANGES_REQUESTED** [review]
Your changes from my previous feedback look great! I had a few more suggestions after looking through this again.

File: src/main/java/org/candlepin/subscriptions/db/model/ServiceLevel.java [hunk 15]
```diff
@@ -33,20 +34,42 @@
...
  * system to associate it with a specific SLA requirement.
  */
-public enum ServiceLevel {
-    UNSPECIFIED(""),
-    PREMIUM("Premium"),
-    STANDARD("Standard"),
-    SELF_SUPPORT("Self-Support"),
-    ANY("_ANY");
+public enum ServiceLevel implements StringValueEnum {
+    EMPTY("") {
```
[the remark this answers is not in the document]
Comment (kahowell) [reply src/main/java/org/candlepin/subscriptions/db/model/ServiceLevel.java:38]: To elaborate, storing it as a field as with `value` would make it much more succinct, as then `asOpenApiEnum()` would simply be something like `return openApiEnum` defined a single time.

File: src/main/java/org/candlepin/subscriptions/db/model/StringValueEnum.java (hunk 12 above)
Comment (awood) [reply src/main/java/org/candlepin/subscriptions/db/model/StringValueEnum.java:34]: :+1: on making the generics a little simpler

File: src/main/java/org/candlepin/subscriptions/db/model/ServiceLevel.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (lindseyburnett) [reply src/main/java/org/candlepin/subscriptions/db/model/ServiceLevel.java:38]: Good call!

File: src/main/java/org/candlepin/subscriptions/db/model/StringValueEnum.java (hunk 11 above)
Comment (lindseyburnett) [reply src/main/java/org/candlepin/subscriptions/db/model/StringValueEnum.java:32]: Thanks for this!  I was trying all kinds of ways to make this more readable; my generics syntax was a little rusty :smile:

File: src/main/java/org/candlepin/subscriptions/db/model/Granularity.java (hunk 4 above)
Comment (lindseyburnett) [reply src/main/java/org/candlepin/subscriptions/db/model/Granularity.java]: After latest refactor, this isn't relevant anymore

**Comment by lindseyburnett** [comment]
@kahowell ready for you again!  I opted not to mess with the static nested Usage.EnumConverter and ServiceLevel.EnumConverter for JPA conversion as part of this - because Generics and all that jazz & Granularity doesn't have one, so I'm thinking it'd have to be another interface or something and not just added to StringValueEnum.

**Comment by lindseyburnett** [comment]
pushed squash

**Review by kahowell — APPROVED** [review]
